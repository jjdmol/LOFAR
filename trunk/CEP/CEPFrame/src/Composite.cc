//#  Composite.cc:
//#
//#  Copyright (C) 2000, 2001
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <CEPFrame/Composite.h>
#include <CEPFrame/NetworkBuilder.h>

namespace LOFAR
{

Composite::Composite()
: Step(),
  itsComposite(0)
{}

Composite::Composite (WorkHolder& worker, 
	      const string& name,
	      bool addNameSuffix,
	      bool controllable,
	      bool monitor)
: Step()
{
  itsComposite = new CompositeRep (worker, name, addNameSuffix, controllable, monitor);
  itsRep = itsComposite;
}

Composite::Composite (WorkHolder* worker, 
	      const string& name,
	      bool addNameSuffix,
	      bool controllable,
	      bool monitor)
: Step()
{
  itsComposite = new CompositeRep (*worker, name, addNameSuffix, controllable, monitor);
  itsRep = itsComposite;
}

Composite::Composite (NetworkBuilder& aBuilder, 
	      const string& name,
	      bool addNameSuffix,
	      bool controllable,
	      bool monitor)
: Step()
{
  itsComposite = new CompositeRep (aBuilder.getWorker(), name, addNameSuffix,
			   controllable, monitor);
  itsRep = itsComposite;
  aBuilder.buildNetwork (*this);
}

Composite::Composite (const Composite& that)
: Step (that)
{
  itsComposite = that.itsComposite;
}

Composite::Composite (CompositeRep* rep)
: Step (rep)
{
  itsComposite = rep;
}

Composite& Composite::operator= (const Composite& that)
{
  if (this != &that) {
    Step::operator= (that);
    itsComposite = that.itsComposite;
  }
  return *this;
}

Composite::~Composite() 
{}


Composite* Composite::clone() const
{
  return new Composite (*this);
}


bool Composite::connectInputArray (Composite* aComposite[],
				   int    nrItems,
				   const TransportHolder& prototype,
				   bool blockingComm)
{
  if (nrItems < 0) {  // set nr_items automatically
    nrItems = getWorker()->getDataManager().getInputs();
  }
  Step** stepPtrs = new Step* [nrItems];
  for (int i=0; i<nrItems; i++) {
    stepPtrs[i] = aComposite[i];
  }
  bool result = itsRep->connectInputArray (stepPtrs, nrItems, prototype, blockingComm);
  delete [] stepPtrs;
  return result;
}

}
