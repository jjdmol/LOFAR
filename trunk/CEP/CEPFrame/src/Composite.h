//# Composite.h: Class to hold a collection of Block objects
//#
//# Copyright (C) 2000, 2001
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_CEPFRAME_COMPOSITE_H
#define LOFAR_CEPFRAME_COMPOSITE_H

//# Never #include <config.h> or #include <lofar_config.h> in a header file!
// \file
// Class to hold a collection of Block objects

#include <CEPFrame/Block.h>
#include <CEPFrame/CompositeRep.h>
//#include <Common/lofar_list.h>

namespace LOFAR
{
// \addtogroup CEPFrame
// @{

class NetworkBuilder;
class VirtualMachine;

/** The Composite class is related to the Block class following a Composite 
    pattern.
    Therefore, a Composite is a collection of Blocks and/or Composites
    The actual simulation work is performed in the process() method
    (which calls its Blocks process() method)
*/

class Composite: public Block
{
public:

  /** Create the Composite with a number inputs and outputs.
      The Composite must get a unique name. To make that process easy,
      by default the suffix _n is added to the name when the Composite is
      added to a Composite (where n is the sequence number starting with 0).
  */
  explicit Composite (int nrInputs,
		      int nrOutputs,
		      const string& name = "aComposite", 
		      bool addNameSuffix = true);

  /// Construct the object using a builder object.
  explicit Composite (int nrInputs,
		      int nrOutputs,
		      NetworkBuilder& builder,
		      const string& name = "aComposite",
		      bool addNameSuffix = true);

  /** Default constructor is only useful for a Composite class member
      (as used in class ApplicationHolder).
      It does not create a useful Composite object.
  */
  Composite();

  /// Copy constructor (reference semantics).
  Composite (const Composite&);

  /// Construct from an existing CompositeRep (meant for internal use).
  explicit Composite (CompositeRep*);

  virtual ~Composite();
  
  /// Assignment (reference semantics).
  Composite& operator= (const Composite&);

  /// Make a correct copy (reference semantics).
  virtual Composite* clone() const;

  /** Add a Block to this Composite
      An exception is thrown if the name of the step (possibly after
      adding the suffix (see constructor)) is not unique.
      Also an exception is thrown if the step has already been added to
      another simul.
  */
  void addBlock (const Block* aBlock)
    { itsComposite->addBlock (*aBlock); }
  void addBlock (const Block& aBlock)
    { itsComposite->addBlock (aBlock); }

  /// Get all Blocks in the Composite.
/*   const list<Block*>& getBlocks() const */
/*     { return itsComposite->getBlocks(); } */

  /// Is this Composite is not part of another Composite?
  bool isHighestLevel() const
    { return itsComposite->isHighestLevel(); }
  /// Clear the highest level flag (thus it is part of another Composite).
  void setNotHighestLevel()
    { itsComposite->setNotHighestLevel(); }

  /// Is the Block a Composite?
  bool isComposite() const; 

  /// Connect this Composite to an array of Composites.
  bool connectInputArray (Composite* aComposite[],   // pointer to  array of ptrs to Blocks
			  int    nrItems, // nr of Blocks in aBlock[] array
			  TransportHolder* prototype,
			  bool blockingComm=true);

  /**
     Set nrChan input channels of aBlock as input channels of this Composite.
     Start at the given indices. 
     If nrChan=-1 is given, it will be set to the 
     minimum number of input channels in aBlock and this Composite.
     This is needed in order to let the framework transport the data read by 
     the Composite directly to the DataHolder of the Block in the Composite.
  */
  void setInput (int thisIndex, 
		 Block* aBlock, 
		 int thatIndex, 
		 int nrChan)
    { return itsComposite->setInput (thisIndex, aBlock, thatIndex, nrChan); }
  
  /**
     Set nrChan output channels of aBlock as output channels of this Composite.
     Start at the given indices.
     If nrChan=-1 is given, it will be set to the minimum number of output
     channels in aBlock and this Composite.
     This is needed in order to let the framework transport the data from 
     the last Block in the Composite to the output of the Composite.
  */
  void setOutput (int thisIndex, 
		  Block* aBlock, 
		  int thatIndex, 
		  int nrChan)
    { return itsComposite->setOutput (thisIndex, aBlock, thatIndex, nrChan); }

  /**
     Set input channels in the aBlock[] array as input channels of 
     this composite.
     This is needed in order to let the framework transport
     the data read by the Composite directly to the DataHolders of the first Blocks
     in the Composite.
     If the nrBlocks argument is -1, it is assumed that
     each of the Blocks in the array has only one output channel
     which will be connected to the input channel of the Composite.
  */
  void setInputArray (Block* aBlock[],  // pointer to  array of ptrs to Blocks
		      int    nrBlocks)  // nr of Blocks in aBlock[] array
    { return itsComposite->setInputArray (aBlock, nrBlocks); }

  /**
     Set output channels in the aBlock[] array as output channels of 
     this composite.
     This is needed in order to let the framework transport
     the data from the last Blocks in the Composite to the output of the Composite.
     If the nrBlocks argument is -1, it is assumed that
     each of the Blocks in the array has only one input channel
     which will be connected to the output channel of the Composite.
   */
  void setOutputArray (Block* aBlock[],  // pointer to  array of ptrs to Blocks
		       int    nrBlocks)  // nr of Blocks in aBlock[] array
    { return itsComposite->setOutputArray (aBlock, nrBlocks); }

  /// Obtain a pointer to the VirtualMachine object
  VirtualMachine& getVM()
    { return itsComposite->getVM(); }

private:
  // Pointer to the actual simul object.
  CompositeRep* itsComposite;
};

inline bool Composite::isComposite() const
  { return true; }

// @}

}
  
#endif
