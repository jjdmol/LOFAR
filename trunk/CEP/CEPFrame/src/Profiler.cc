//  Profiler.cc:
//
//  Copyright (C) 2000, 2001
//  ASTRON (Netherlands Foundation for Research in Astronomy)
//  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//  $Id$
//
//
/////////////////////////////////////////////////////////////////////////////////////

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "CEPFrame/Profiler.h"
#include "Common/Debug.h"
#include <Common/lofar_iostream.h>

#ifdef HAVE_MPI_PROFILER

int  Profiler::theirNextFreeState=1001;
bool Profiler::theirIsActive=false;    // start in de-activate state

#include "mpi.h"

#ifdef HAVE_MPI
#include "/opt/scali/contrib/mpe/include/mpe.h"
#endif

void Profiler::init() {
  MPE_Init_log();
  MPI_Pcontrol(0);        // switch off MPI build-in profiling
  theirIsActive = false;   // start in de-activate state
}

int Profiler::defineState (const char* name, const char* color)
{
  int startstate = theirNextFreeState++; // the start state
  int endstate   = theirNextFreeState++; // the end state
  MPE_Describe_state (startstate, endstate, (char*)name, (char*)color);
  cdebug(3) << "Defined State " << name
	    << " start=" << startstate
	    << "  end=" << endstate << endl;
  return startstate;
}

void Profiler::enterState (int astate)
{
  if (theirIsActive) {
    MPE_Log_event (astate, 1, (char*)0);
  }
  return;
}

void Profiler::leaveState (int astate)
{
  if (theirIsActive) {
    MPE_Log_event (astate+1, 1, (char*)0);
  }
  return;
}


void Profiler::activate()
{
  MPE_Start_log();
  // switch on the Profiler
  MPI_Pcontrol (1);         // switch on MPI build-in profiling
  theirIsActive = true;
}


void Profiler::deActivate()
{
  MPE_Finish_log("MPIlog");
  // switch off the Profiler
  MPI_Pcontrol (0);         // switch off MPI build-in profiling
  theirIsActive = false;
}


#endif
