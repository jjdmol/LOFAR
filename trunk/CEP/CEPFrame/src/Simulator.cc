//  Simulator.cc:
//
//  Copyright (C) 2000, 2001
//  ASTRON (Netherlands Foundation for Research in Astronomy)
//  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//  $Id$
//
/////////////////////////////////////////////////////////////////////////////

#include "CEPFrame/Simulator.h"
#include "CEPFrame/SimulatorParseClass.h"
#include TRANSPORTERINCLUDE

Simulator::Simulator()
: itsSimul(),
  itsArgc (0),
  itsArgv (0),
  itsPreDone (true),
  itsPostDone(true)
{}

Simulator::~Simulator()
{}

void Simulator::setarg (int argc, const char** argv)
{
  itsArgc = argc;
  itsArgv = argv;
}

void Simulator::getarg (int* argc, const char** argv[])
{
  *argc = itsArgc;
  *argv = itsArgv;
}

void Simulator::baseDefine (const ParamBlock& params)
{
  // Initialize MPI environment.
  TRANSPORTER::init (itsArgc, itsArgv);
  // Set current application number if defined in parameters.
  ParamBlock::const_iterator iter = params.find ("appl");
  if (iter != params.end()) {
    if (iter->second.dataType() == ParamValue::DTInt) {
      Step::setCurAppl (iter->second.getInt());
    }
  }
  // Let derived class define the simulation.
  define (params);
  itsPreDone = false;
  /////  itsSimul.shortcutConnections();
  /////  itsSimul.simplifyConnections();
}

void Simulator::baseCheck()
{
  itsSimul.checkConnections();
  check();
}

void Simulator::basePrerun()
{
  prerun();
  itsPreDone = true;
  itsPostDone = false;
}

void Simulator::baseRun (int nsteps)
{
  if (!itsPreDone) {
    basePrerun();
  }
  run (nsteps);
}

void Simulator::basePostrun()
{
  if (!itsPreDone) {
    basePrerun();
  }
  postrun();
  itsPostDone = true;
  itsPreDone = false;
}

void Simulator::baseDump()
{
  if (!itsPreDone) {
    basePrerun();
  }
  dump();
}

void Simulator::baseDHFile (const string& dh, const string& name)
{
  if (!itsPreDone) {
    basePrerun();
  }
  itsSimul.setDHFile (dh, name);
}

void Simulator::baseQuit()
{
  if (!itsPostDone) {
    basePostrun();
  }
  quit();
  // Close environment.
  TRANSPORTER::finalize();
}

void Simulator::check()
{}

void Simulator::prerun()
{
  itsSimul.preprocess();
}

void Simulator::dump() const
{}

void Simulator::postrun()
{
  itsSimul.postprocess();
}

void Simulator::quit()
{}
