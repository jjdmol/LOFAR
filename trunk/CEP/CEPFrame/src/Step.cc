//#  Step.cc:
//#
//#  Copyright (C) 2000, 2001
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <CEPFrame/Step.h>
#include <CEPFrame/Composite.h>

namespace LOFAR
{

Step::Step (WorkHolder& worker, 
	    const string& aName,
	    bool addNameSuffix,
	    bool monitor)
: itsRep(0)
{
  itsRep = new StepRep (worker, aName, addNameSuffix, monitor);
}

Step::Step (WorkHolder* worker, 
	    const string& aName,
	    bool addNameSuffix,
	    bool monitor)
: itsRep(0)
{
  itsRep = new StepRep (*worker, aName, addNameSuffix, monitor);
}

Step::Step (const Step& that)
{
  itsRep = that.itsRep;
  if (itsRep) {
    itsRep->incrRefCount();
  }
}

Step::Step (StepRep* rep)
{
  itsRep = rep;
  if (itsRep) {
    itsRep->incrRefCount();
  }
}

Step& Step::operator= (const Step& that)
{
  if (this != &that) {
    if (itsRep  &&  itsRep->decrRefCount() <= 0) {
      delete itsRep;
    }
    itsRep = that.itsRep;
    if (itsRep) {
      itsRep->incrRefCount();
    }
  }
  return *this;
}

Step::~Step() 
{
  if (itsRep  &&  itsRep->decrRefCount() <= 0) {
    delete itsRep;
  }
}

Step* Step::clone() const
{
  return new Step (*this);
}

Composite Step::getParent() const
{
  return Composite(itsRep->getParent());
}

}
