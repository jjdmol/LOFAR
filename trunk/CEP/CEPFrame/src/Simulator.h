//  Simulator.h: Abstract base class for simulator programs
//
//  Copyright (C) 2000, 2001
//  ASTRON (Netherlands Foundation for Research in Astronomy)
//  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//  $Id$
//
/////////////////////////////////////////////////////////////////////////////////

#ifndef CEPFRAME_SIMULATOR_H
#define CEPFRAME_SIMULATOR_H

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "CEPFrame/Simulator.h"
#include "CEPFrame/Simul.h"
#include "CEPFrame/ParamBlock.h"


/**
   Simulator is the abstract base class for simulation application programs.
   It defines a few methods that need to be implemented in a derived class
   doing the simulation.

   The main functions that have to be implemented are define and run.
   Define should set up the simulation. Note that the base class already
   does the initialization and finalization of the transporter.
*/

class Simulator
{
public:
  /// Default constructor does nothing.
  Simulator();

  virtual ~Simulator();

  /// Set the command line argument pointers (to be used by MPI).
  void setarg (int argc, const char* argv[]);

  /// Get the command line argument pointers (to be used by MPI).
  void getarg (int* argc, const char** argv[]);

  /** Define the simulation.
      It starts the transport and calls the define function.
      At the end it resolves the communication.
  */
  void baseDefine (const ParamBlock& params = ParamBlock());

  /** Check the the simulation.
      The connections are checked and thereafter the user check()
      function is called.
  */
  void baseCheck();

  /// Do a prerun on the simulation.
  void basePrerun();

  /// Run the simulation by calling the run function.
  void baseRun (int nsteps = -1);

  /// Dump the simulation data by calling the dump function.
  void baseDump();

  /// Set the output file of a data holder.
  void baseDHFile (const string& dh, const string& name);

  /// Do a postrun on the simulation.
  void basePostrun();

  /** Quit the simulation.
      It calls the quit function and closes the transport.
  */
  void baseQuit();

protected:
  /**@name Virtual functions
     @memo Functions to be implemented in derived class.
     @doc The following functions can or have to be implemented in
     a derived class. The default implementations of the non-pure virtual
     functions do nothing.
     These virtual functions are called by their base... counterparts.
  */
  //@{
  virtual void define (const ParamBlock&) = 0;
  virtual void check();
  virtual void prerun();
  virtual void run (int nsteps) = 0;
  virtual void dump() const;
  virtual void postrun();
  virtual void quit();
  //@}

  /// Fill the pointer to the simulation.
  void setSimul (const Simul&);

  /// Get the pointer to the simulation.
  Simul& getSimul();
  const Simul& getSimul() const;

private:
  // Forbid copy constructor.
  Simulator (const Simulator&);

  // Forbid assignment.
  Simulator& operator= (const Simulator&);


  /** A pointer to the simulation.
      Note that the base destructor does not delete the pointer.
  */
  Simul itsSimul;

  int    itsArgc;
  const char** itsArgv;
  bool   itsPreDone;
  bool   itsPostDone;
};


inline void Simulator::setSimul (const Simul& simul)
  { itsSimul = simul; }

inline Simul& Simulator::getSimul()
  { return itsSimul; }

inline const Simul& Simulator::getSimul() const
  { return itsSimul; }



#endif
