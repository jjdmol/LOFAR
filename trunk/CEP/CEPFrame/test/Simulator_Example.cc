//  Simulator_Example.cc: Program for testing simulation parser
//
//  Copyright (C) 2000-2002
//  ASTRON (Netherlands Foundation for Research in Astronomy)
//  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//  $Id$
//
/////////////////////////////////////////////////////////////////////////

#include <lofar_config.h>

#include <Simulator_Example.h>
#include <tinyCEP/SimulatorParseClass.h>
#include <CEPFrame/Step.h>
#include <Transport/TH_Mem.h>
#include <CEPFrame/Composite.h>
#include <CEPFrame/WH_Example.h>
#include <tinyCEP/Profiler.h>
#include <Common/KeyValueMap.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_iostream.h>

using namespace LOFAR;

Simulator_Example::~Simulator_Example()
{}

void Simulator_Example::define (const KeyValueMap& params)
{
  params.show (cout);

  // define the top-level composite object
  WH_Example exSimul("ExampleSimul");
  Composite comp(exSimul);
  setComposite (comp);

  // tell the Composite where to run
  comp.runOnNode(0);
  
  // Now start filling the simulation. 
  // first create the Steps
  WH_Example whEx1("Step1");
  Step step1(whEx1);
  WH_Example whEx2("Step2");
  Step step2(whEx2);
  WH_Example whEx3("Step3");
  Step step3(whEx3);

  // Create the first composite and fill it
  WH_Example whComp1("Composite1");
  Composite composite1(whComp1, "simcomp1", false);
  composite1.addStep(step2);
  composite1.addStep(step3);

  // Create the second composite and fill it
  WH_Example whComp2("Composite2");
  Composite composite2(whComp2, "simcomp2", false);
  composite2.addStep(step1);
  composite2.addStep(composite1);

  // Add the Composite2 to the top-level Composite
  comp.addStep(composite2);

  // Now define the connections between the Steps and Composite objects:
  step3.connectInput(&step2,TH_Mem(), false);
  Step* addr2 = &composite2;
  comp.connectInputToArray(&addr2,1,0,0,TH_Mem(),false);

  //////////////////////////////////////////////////////////////////////
  //
  // Finished configuration definition
  // 
  //////////////////////////////////////////////////////////////////////
}

void Simulator_Example::run (int nsteps)
{
  if (nsteps < 0) {
    nsteps = 10;
  }
  LOG_TRACE_FLOW("Ready with definition of configuration");
  Profiler::init();

  cout << endl <<  "Start Process" << endl;    
  for (int i=0; i<nsteps; i++) {
    if (i==2) Profiler::activate();
    getComposite().process();
    if (i==5) Profiler::deActivate();
  }
}

void Simulator_Example::dump() const
{

  cout << endl << "DUMP Data from last Processing step: " << endl;
  getComposite().dump();
}

void Simulator_Example::quit()
{

}



int main (int argc, const char* argv[])
{
  // Set trace level.
  INIT_LOGGER("SimulatorExample.log_prop");
  try {
    // First test some Parse functions.
    // Note that C++ also requires a \ to escape special characters.
    ASSERTSTR (SimulatorParse::removeQuotes ("'abcd\"e'\"fgh'i\"'j'")
	       == string("abcd\"efgh'ij"),
	       "SimulatorParse::removeQuotes fails");
    ASSERTSTR (SimulatorParse::removeEscapes ("\\ab\\\\cd\\efghij\\")
	       == string("ab\\cdefghij"),
	       "SimulatorParse::removeEscapes fails");

    Simulator_Example simulator;
    simulator.setarg (argc, argv);
// #ifndef HAVE_MPI
    cout << endl;
    cout << "  * Type 'define;' to define the simulation" << endl;
    cout << "  * Type 'run;'    to run the simulation" << endl;
    cout << "  * Type 'dump;'   to dump the simulator data" << endl;
    cout << "  * Type 'quit'    to quit" << endl;
    cout << endl;
    try {
      SimulatorParse::parse (simulator);
    } catch (SimulatorParseError& x) {
      cout << x.what() << endl;
    }
    cout << endl;
    cout << "It was a pleasure working with you!" << endl << endl;
// #else
//     cout << "Welcome to LOFARSim" <<endl;
//     cout << "Running in batch mode " << endl;
//     cout << endl;
//     cout << "Call define" << endl;
//     simulator.baseDefine();
//     cout << endl;
//     cout << "Call run" << endl;
//     simulator.baseRun();
//     cout << endl;
//     cout << "Call dump " << endl;
//     simulator.baseDump();
//     cout << "Call quit " << endl;
//     simulator.baseQuit();
//     cout << endl;
//     cout << "Good Bye!" << endl;
// #endif
  } catch (...) {
    cout << "Unexpected exception";
  }
}
