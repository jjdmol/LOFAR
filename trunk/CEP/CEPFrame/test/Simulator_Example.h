//  Simulator_Example.h: Example of a concrete Simulator class
//
//  Copyright (C) 2000, 2001
//  ASTRON (Netherlands Foundation for Research in Astronomy)
//  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//  $Id$
//
//  $Log$
//  Revision 1.3  2002/05/07 08:54:36  gvd
//  Added package to includes
//
//  Revision 1.2  2001/10/26 10:06:28  wierenga
//  Wide spread changes to convert from Makedefs to autoconf/automake/libtool build environment
//
//  Revision 1.1  2001/08/30 09:12:23  gvd
//  moved to test dir.
//
//  Revision 1.4  2001/03/23 10:00:40  gvd
//  Improved documentation and test programs
//  Added clearEventCount function to Step
//
//  Revision 1.3  2001/03/01 13:15:47  gvd
//  Added type argument in DataHolder constructor which is used in
//  the connect functions to check if the DH types match
//  Improved the simulator parser
//  Improved documentation
//
//  Revision 1.2  2001/02/05 14:53:05  loose
//  Added GPL headers
//
//////////////////////////////////////////////////////////////////////////

#ifndef BASESIM_SIMULATOR_EXAMPLE_H
#define BASESIM_SIMULATOR_EXAMPLE_H

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "BaseSim/Simulator.h"

/**
   This class is an example of a concrete Simulator.
*/

class Simulator_Example: public Simulator
{
public:
  virtual ~Simulator_Example();

  virtual void define (const ParamBlock&);
  virtual void run (int nsteps);
  virtual void dump() const;
  virtual void quit();
};


#endif
