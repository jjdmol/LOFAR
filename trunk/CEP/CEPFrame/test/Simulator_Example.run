#!/bin/sh

./Simulator_Example -dGlobal=0 < Simulator_Example.in
STATUS=$?
exit $STATUS
