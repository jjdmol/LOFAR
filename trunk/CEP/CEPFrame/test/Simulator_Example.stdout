
  * Type 'define;' to define the simulation
  * Type 'run;'    to run the simulation
  * Type 'dump;'   to dump the simulator data
  * Type 'quit'    to quit


Start Process
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)
Step1
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step2
Processing Data! (0,0),(0,0)   --->  (1,0),(1,0)
Step3
Processing Data! (0,0),(0,0)   --->  (2,0),(2,0)

DUMP Data from last Processing step: 
WH_Example Step1 Buffers:
(1,0),(1,0)
WH_Example Step2 Buffers:
(1,0),(1,0)
WH_Example Step3 Buffers:
(2,0),(2,0)

It was a pleasure working with you!

